tilequeue/rawr.py

```diff
@@ -0,0 +1,254 @@
+        except Exception as e:
+            # boto3 client treats 304 responses as exceptions
+            if isinstance(e, ClientError):
+                resp = getattr(e, 'response', None)
+                assert resp
+            else:
+                raise e
+        status_code = resp['ResponseMetadata']['HTTPStatusCode']
+        if status_code == 304:
+            assert self.prev_toi
+            toi = self.prev_toi
+        elif status_code == 200:
+            body = resp['Body']
+            try:
+                gzip_payload = body.read()
+            finally:
+                try:
+                    body.close()
+                except:
+                    pass
+            gzip_file_obj = StringIO(gzip_payload)
+            toi = load_set_from_gzipped_fp(gzip_file_obj)
+            self.prev_toi = toi
+            self.etag = resp['ETag']
+        else:
+            assert 0, 'Unknown status code from toi get: %s' % status_code
+
+        return toi
+
+    def __call__(self, coords):
+        toi = self.tiles_of_interest()
+        coord_ints = convert_to_coord_ints(coords)
+        intersected_coord_ints, _ = explode_and_intersect(coord_ints, toi)
+        for coord_int in intersected_coord_ints:
+            coord = coord_unmarshall_int(coord_int)
+            yield coord
+
+
+class RawrTileGenerationPipeline(object):
+
+    """Entry point for rawr process command"""
+
+    def __init__(
+            self, rawr_queue, msg_marshaller, group_by_zoom, rawr_gen,
+            queue_writer, rawr_toi_intersector):
+        self.rawr_queue = rawr_queue
+        self.msg_marshaller = msg_marshaller
+        self.group_by_zoom = group_by_zoom
+        self.rawr_gen = rawr_gen
+        self.queue_writer = queue_writer
+        self.rawr_toi_intersector = rawr_toi_intersector
+
+    def __call__(self):
+        # TODO - should probably provide a way to stop iteration
+        while True:
+            msg_handle = self.rawr_queue.read()
+            coords = self.msg_marshaller.unmarshall(msg_handle.payload)
+            parent = common_parent(coords, self.group_by_zoom)
+            rawr_tile_coord = convert_coord_object(parent)
+            self.rawr_gen(rawr_tile_coord)
+            coords_to_enqueue = self.rawr_toi_intersector(coords)
+            self.queue_writer.enqueue_batch(coords_to_enqueue)
+            self.rawr_queue.done(msg_handle)
+
+
+def make_rawr_zip_payload(rawr_tile, date_time=None):
+    """make a zip file from the rawr tile formatted data"""
+    if date_time is None:
+        date_time = gmtime()[0:6]
+
+    buf = StringIO()
+    with zipfile.ZipFile(buf, mode='w') as z:
+        for fmt_data in rawr_tile.all_formatted_data:
+            zip_info = zipfile.ZipInfo(fmt_data.name, date_time)
+            z.writestr(zip_info, fmt_data.data, zipfile.ZIP_DEFLATED)
+    return buf.getvalue()
+
+
+def make_s3_path(tile, prefix):
+    return '%s/%d/%d/%d.zip' % (prefix, tile.z, tile.x, tile.y)
```

Comment:
Should we be hash-prefixing the tiles? I guess there's only a million of them, but it would be theoretically nice to be able to write all million in parallel (which means having at least 3 hex digits of hash - 5,000 partitions).

Reply:
Probably, good catch!

I added that in cb6ad62. I also had it take a suffix for the extension, which most likely doesn't make a difference practically, but it's how everything else is stored too.